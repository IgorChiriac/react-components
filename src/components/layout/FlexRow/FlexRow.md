The `FlexRow` component is usually used as a child of `FlexContainer`s.
However, in case of nested grids it can also be a child of a `FlexCol` or any div element.
`FlexCol`s should be his direct children.
