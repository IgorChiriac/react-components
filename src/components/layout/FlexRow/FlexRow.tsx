import React from 'react';
import styled from 'styled-components';
import grid from '../../../constants/grid';

export type TFlexValues = 'flex-start' | 'center' | 'flex-end' | 'space-between' | 'space-around' | 'space-evenly';

export interface IFlexRowProps {
  align?: TFlexValues;
  cols?: number;
  gutter?: number;
  justify?: TFlexValues;
}

export interface IFlexRow extends React.HTMLAttributes<HTMLDivElement>, IFlexRowProps {}

const defaultProps: Partial<IFlexRow> = {
  align: 'flex-start',
  cols: grid.cols,
  gutter: grid.gutter,
  justify: 'flex-start',
};

const StyledFlexRow = styled.div<IFlexRow>`
  display: flex;
  flex-wrap: wrap;
  margin-right: -${props => props.gutter}px;
  margin-left: -${props => props.gutter}px;
  justify-content: ${props => props.justify};
  align-items: ${props => props.align};
`;

const FlexRow: React.FunctionComponent<IFlexRow> = ({ children, ...props }) => {
  const childrenWithProps = React.Children.toArray(children)
    .filter(child => !!child)
    .map(child => React.cloneElement(child as any, { gutter: props.gutter, cols: props.cols }));
  return <StyledFlexRow {...props}>{childrenWithProps}</StyledFlexRow>;
};

FlexRow.defaultProps = defaultProps;

export default FlexRow;
