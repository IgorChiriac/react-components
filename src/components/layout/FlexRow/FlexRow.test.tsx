import React from 'react';
import { render } from 'react-testing-library';
import FlexRow from './FlexRow';

describe('<FlexRow />', () => {
  it('renders a FlexRow component', () => {
    const { container } = render(<FlexRow />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
