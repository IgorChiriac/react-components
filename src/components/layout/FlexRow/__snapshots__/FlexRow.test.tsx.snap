// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlexRow /> renders a FlexRow component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-right: -16px;
  margin-left: -16px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

<div
  class="c0"
  cols="12"
/>
`;
